agent/checks/check_test.go

```diff
@@ -936,6 +937,115 @@ func TestCheckTCPPassing(t *testing.T) {
 	tcpServer.Close()
 }
 
+func TestCheckH2PING(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		desc        string
+		passing     bool
+		timeout     time.Duration
+		connTimeout time.Duration
+	}{
+		{desc: "passing", passing: true, timeout: 1 * time.Second, connTimeout: 1 * time.Second},
+		{desc: "failing because of time out", passing: false, timeout: 1 * time.Nanosecond, connTimeout: 1 * time.Second},
+		{desc: "failing because of closed connection", passing: false, timeout: 1 * time.Nanosecond, connTimeout: 1 * time.Millisecond},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { return })
+			server := httptest.NewUnstartedServer(handler)
+			server.EnableHTTP2 = true
+			server.Config.ReadTimeout = tt.connTimeout
+			server.StartTLS()
+			defer server.Close()
+			serverAddress := server.Listener.Addr()
+			target := serverAddress.String()
+
+			notif := mock.NewNotify()
+			logger := testutil.Logger(t)
+			statusHandler := NewStatusHandler(notif, logger, 0, 0)
+			cid := structs.NewCheckID("foo", nil)
+			tlsCfg := &api.TLSConfig{
+				InsecureSkipVerify: true,
+			}
+			tlsClientCfg, err := api.SetupTLSConfig(tlsCfg)
+			if err != nil {
+				t.Fatalf("%v", err)
+			}
+			tlsClientCfg.NextProtos = []string{http2.NextProtoTLS}
+
+			check := &CheckH2PING{
+				CheckID:         cid,
+				H2PING:          target,
+				Interval:        5 * time.Second,
+				Timeout:         tt.timeout,
+				Logger:          logger,
+				TLSClientConfig: tlsClientCfg,
+				StatusHandler:   statusHandler,
+			}
+
+			check.Start()
+			defer check.Stop()
+
+			if tt.passing == true {
+				retry.Run(t, func(r *retry.R) {
+					if got, want := notif.State(cid), api.HealthPassing; got != want {
+						r.Fatalf("got state %q want %q", got, want)
+					}
+				})
+			} else {
+				retry.Run(t, func(r *retry.R) {
+					if got, want := notif.State(cid), api.HealthCritical; got != want {
+						r.Fatalf("got state %q want %q", got, want)
+					}
+				})
+			}
+		})
+	}
+}
+
+func TestCheckH2PINGInvalidListener(t *testing.T) {
```

Comment:
An additional test case, similar to [https://github.com/hashicorp/consul/blob/bbc69636ec0553cc273692fd9cda07d1e3308987/agent/checks/check_test.go#L730-L770](TestCheckHTTP_TLS_BadVerify) would be good to verify that the correct TLS certs are being used.